Sources/Public/Internal/CrashLoopDetector/BacktraceCrashLoopDetector.swift

```diff
@@ -0,0 +1,180 @@
+         plus they allow to avoid a headache with reading/writing to/from the custom file.
+
+         - But we should consider shared computers as well - comment from UserDefaults docs:
+            With the exception of managed devices in educational institutions,
+            a user’s defaults are stored locally on a single device,
+            and persisted for backup and restore.
+            To synchronize preferences and other data across a user’s connected devices,
+            use NSUbiquitousKeyValueStore instead.
+         */
+        guard let data = UserDefaults.standard.object(forKey: BacktraceCrashLoopDetector.plistKey) as? Data
+        else { return }
+
+        guard let array = try? PropertyListDecoder().decode([StartUpEvent].self, from: data)
+        else { return }
+
+        startupEvents.append(contentsOf: array)
+        BacktraceCrashLoopDetector.LogDebug("Events Loaded: \(startupEvents.count)")
+    }
+
+    @objc internal func saveEvents() {
+        let data = try? PropertyListEncoder().encode(startupEvents)
+        UserDefaults.standard.set(data, forKey: BacktraceCrashLoopDetector.plistKey)
+        BacktraceCrashLoopDetector.LogDebug("Events Saved: \(startupEvents.count)")
+    }
+
+    @objc internal func addStartupEvent() {
+        
+        let crashesCount = BacktraceCrashLoopCounter.crashesCount()
+        let event = StartUpEvent(uuid: UUID().uuidString,
+                                 timestamp: Double(Date.timeIntervalSinceReferenceDate),
+                                 crashesCount: crashesCount)
+        
+        BacktraceCrashLoop.LogDebug(event.description())
+
+        startupEvents.append(event)
+        
+        if startupEvents.count > BacktraceCrashLoopDetector.consecutiveCrashesThreshold {
+            startupEvents.removeFirst()
+        }
+
+        BacktraceCrashLoop.LogDebug("Startup Event Added: \(startupEvents.count)")
+    }
+
+    @objc internal func clearStartupEvents() {
+        startupEvents.removeAll()
+        saveEvents()
+        BacktraceCrashLoop.LogDebug("Startup Events Cleared: \(startupEvents.count)")
+    }
+    
+    @objc internal func consecutiveEventsCount() -> Int {
+        
+        var count = 0
+        var previousValue = 0
+        for event in startupEvents {
+            if event.crashesCount > previousValue {
+                count += 1
+            }
+            previousValue = event.crashesCount
+        }
+
+        BacktraceCrashLoop.LogDebug("Consecutive Events Count: \(count)")
+        return count
+    }
+    
+    @objc internal func databaseDescription() -> String {
+        var string = ""
+        for event in startupEvents {
+            string += event.description() + "\n"
+        }
+        return string
+    }
+}
+
+// MARK: Deprecated methods
+extension BacktraceCrashLoopDetector {
+
+    @available(*, deprecated, message: "Temporarily not needed")
+    @objc internal func reportFilePath() -> String {
```

Comment:
have you consider a situation when the crash report is from the previous session because we couldn't send it? To reproduce this problem disable internet on your phone and restart application twice. Would we increase the counter?

Reply:
That's why we have a counter, counter incremented every time crash happens and compared with previous value from user defaults upon next launch.

Reply:
counter increase only when crash happens, not when launch happens.
But - I'm reconsidering idea of counter - if we'll be checking crash file creation date at launch - counter wont be necessary.

Reply:
In new logic - if file has same timestamp - it won't be considered as a crash loop continuation.